query/workmanager.go

```diff
@@ -0,0 +1,464 @@
+
+	// When the work dispatcher exits, we'll loop through the remaining
+	// batches and send on their error channel.
+	defer func() {
+		for _, b := range currentBatches {
+			b.errChan <- ErrWorkManagerShuttingDown
+		}
+	}()
+
+	// We set up a counter that we'll increase with each incoming query,
+	// and will serve as the priority of each. In addition we map each
+	// query to the batch they are part of.
+	queryIndex := uint64(0)
+	currentQueries := make(map[uint64]uint64)
+
+	workers := make(map[string]*activeWorker)
+
+Loop:
+	for {
+		// If the work queue is non-empty, we'll take out the first
+		// element in order to distribute it to a worker.
+		if work.Len() > 0 {
+			next := work.Peek().(*queryJob)
+
+			// Find the peers with free work slots available.
+			var freeWorkers []string
+			for p, r := range workers {
+				// Only one active job at a time is currently
+				// supported.
+				if r.activeJob != nil {
+					continue
+				}
+
+				freeWorkers = append(freeWorkers, p)
+			}
+
+			// Use the historical data to rank them.
+			w.cfg.Ranking.Order(freeWorkers)
+
+			// Give the job to the highest ranked peer with free
+			// slots available.
+			for _, p := range freeWorkers {
+				r := workers[p]
+
+				// The worker has free work slots, it should
+				// pick up the query.
+				select {
+				case r.w.NewJob() <- next:
+					log.Tracef("Sent job %v to worker %v",
+						next.Index(), p)
+					heap.Pop(work)
+					r.activeJob = next
+
+					// Go back to start of loop, to check
+					// if there are more jobs to
+					// distribute.
+					continue Loop
+
+				// Remove workers no longer active.
+				case <-r.onExit:
+					delete(workers, p)
+					continue
+
+				case <-w.quit:
+					return
+				}
+			}
+		}
+
+		// Otherwise the work queue is empty, so we'll just wait for a
+		// result of a previous query to come back, a new peer to
+		// connect, or for a new batch of queries to be scheduled.
+		select {
+
+		// Spin up a goroutine that runs a worker each time a peer
+		// connects.
+		case peer := <-peersConnected:
+			log.Debugf("Starting worker for peer %v",
+				peer.Addr())
+
+			r := w.cfg.NewWorker(peer)
+
+			// We'll create a channel that will close after the
+			// worker's Run method returns, to know when we can
+			// remove it from our set of active workers.
+			onExit := make(chan struct{})
+			workers[peer.Addr()] = &activeWorker{
+				w:         r,
+				activeJob: nil,
+				onExit:    onExit,
+			}
+
+			w.cfg.Ranking.AddPeer(peer.Addr())
+
+			w.wg.Add(1)
+			go func() {
+				defer w.wg.Done()
+				defer close(onExit)
+
+				r.Run(w.jobResults, w.quit)
+			}()
+
+		// A new result came back.
+		case result := <-w.jobResults:
+			log.Tracef("Result for job %v received from peer %v "+
+				"(err=%v)", result.job.index,
+				result.peer.Addr(), result.err)
+
+			// Delete the job from the worker's active job, such
+			// that the slot gets opened for more work.
+			r, ok := workers[result.peer.Addr()]
+			if ok {
```

Comment:
Why is this check needed?

Reply:
I don't think it is! Removed: https://github.com/lightninglabs/neutrino/pull/179/commits/43dd217aa8438064fe29febc707432253a8a7fcc